packages/components/paginator/src/paginator.ts

```diff
@@ -49,489 +57,295 @@ export class Paginator extends ScopedElementsMixin(LitElement) {
+        aria-current=${ifDefined(this.page === this.pageCount - 1 ? 'page' : undefined)}
+        class=${classMap({ current: this.page === this.pageCount - 1, page: true })}
+        fill="ghost"
+        size="lg"
+      >
+        ${this.pageCount}
+      </sl-button>
+
+      <div class="wrapper">
+        <sl-select
+          @sl-change=${this.#onChange}
+          .value=${this.page}
+          aria-label=${`${msg(str`${this.page}, page`)}`}
           size="lg"
-          ?disabled=${this.page === this.#pages}
-          @click=${this.#onNext}
         >
-          <sl-icon name="caret-right-solid"></sl-icon>
-        </sl-button>
-      </nav>
+          ${Array.from({ length: this.pageCount }).map(
+            (_, index) => html`
+              <sl-select-option aria-label=${msg(str`${index + 1}, page`)} .value=${index}>
+                ${index + 1}
+              </sl-select-option>
+            `
+          )}
+        </sl-select>
+        <span>of ${this.pageCount} pages</span>
+      </div>
+
+      <sl-button
+        @click=${this.#onNext}
+        ?disabled=${this.page === this.pageCount - 1}
+        aria-label=${msg(str`Go to the next page (${this.page + 2})`)}
+        class="nav"
+        fill="ghost"
+        size="lg"
+      >
+        <sl-icon name="caret-right-solid"></sl-icon>
+      </sl-button>
     `;
   }
 
-  /** Handles `click` event on the previous button. */
-  #onPrevious() {
-    this.page--;
-
-    if (this.dataSource) {
-      this.dataSource.setPage(this.page);
-      this.dataSource.update();
-    }
-
-    this.#setCurrentlyVisibleItems();
-    this.#updateVisibility();
+  #onChange(event: SlChangeEvent<number>): void {
+    this.#onPageClick(event.detail);
   }
 
-  /** Handles `click` event on the next button. */
   #onNext() {
-    this.page++;
-
-    if (this.dataSource) {
-      this.dataSource.setPage(this.page);
-      this.dataSource.update();
-    }
-
-    this.#setCurrentlyVisibleItems();
-    this.#updateVisibility();
+    this.#onPageClick(Math.min(this.page + 1, this.pageCount - 1), true);
   }
 
-  #setActive(event: Event) {
-    const target = event.target as Select | PaginatorPage;
-
-    if (target instanceof Select) {
-      this.page = Number(target.value);
-    } else {
-      this.page = Number(target.innerText?.trim());
-    }
+  #onPageClick(page: number, announcePage = false): void {
+    this.page = page;
+    this.pageChangeEvent.emit(this.page);
 
     if (this.dataSource) {
       this.dataSource.setPage(this.page);
       this.dataSource.update();
     }
 
-    this.#setCurrentlyVisibleItems();
     this.#updateVisibility();
 
-    requestAnimationFrame(() => {
-      const activePageElement = this.renderRoot.querySelector('[active]') as PaginatorPage;
-      const activeButton = activePageElement?.renderRoot.querySelector('button');
-      activeButton?.focus();
-    });
+    if (announcePage) {
+      announce(msg(str`Page ${this.page} of ${this.pageCount}`));
```

Comment:
Page change should be also announced when we change page size and the page is being set to `1`.

Reply:
Fixed